### PR TITLE
 refactor(api): limit gen2 mag module max engage height

### LINKED ISSUE #5212: api: change magnetic module motion extents for gen2
Gen2 magnetic modules will correctly limit their motion to +25mm rather than +45; this limit should be enforced in the hardware controller and in the protocol API.

### PR DESCRIPTION
**Pull request by ahiuchingau**
## overview
This PR limits the Gen2 Magnetic Module engage height to 25 mm at 100 steps/mm.

In order to enforce this limit even during simulation, the simulating driver would need to know which model was loaded in the protocol. Therefore, the hardware module will now take an closes #5212
 argument `sim_model` during simulation from protocol context, which will then be evaluated to determine the correct max engage height. 

Note: Gen1 magnetic module behavior remains unchanged.
closes #5212

## changelog
- passed module model 
- added and updated appropriate tests

## review requests
To test this PR irl, you will first need to add the changes in the Gen2 Mag Module firmware in the this PR - https://github.com/Opentrons/opentrons-modules/pull/114

- check that during both run and simulation, engage height > 25 mm should raise an error for Gen2 Magdeck
- check that max engage height for Gen1 is still 45 mm

## risk assessment
Low risk because gen1's behavior should be the same

### CODE REVIEW

File: api/src/opentrons/hardware_control/modules/tempdeck.py [hunk 1]
```diff
@@ -152,7 +153,8 @@ def __init__(self,
                  port: str,
                  execution_manager: ExecutionManager,
                  simulating: bool,
-                 loop: asyncio.AbstractEventLoop = None) -> None:
+                 loop: asyncio.AbstractEventLoop = None,
+                 sim_model: str = None) -> None:
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/modules/tempdeck.py]: same here re: push down to the driver

File: api/src/opentrons/hardware_control/modules/thermocycler.py [hunk 2]
```diff
@@ -161,7 +161,8 @@ def __init__(self,
                  execution_manager: ExecutionManager,
                  interrupt_callback: types.InterruptCallback = None,
                  simulating: bool = False,
-                 loop: asyncio.AbstractEventLoop = None) -> None:
+                 loop: asyncio.AbstractEventLoop = None,
+                 sim_model: str = None) -> None:
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/modules/thermocycler.py]: if we're going to add this to all three we should put it in the abstract base class

**Review by sfoster1 — CHANGES_REQUESTED** [review]
Couple small changes.

I'm also not really sure about merging this until we make the change in the module firmware - if we merge this now, gen2 magdecks will become unusable until that is merged.

**Review by sfoster1 — CHANGES_REQUESTED** [review]
Let's also move the simulating drivers to `opentrons.drivers.{module_name}`, ahve the modules connect to the simulators, get an appropriate hw revision based on the sim model, and have a single builder function in `opentrons.hardware_control.modules` to build the simulated modules.

File: api/src/opentrons/hardware_control/modules/tempdeck.py [hunk 3]
```diff
@@ -142,26 +90,30 @@ def bootloader(cls) -> mod_abc.UploadFunction:
...
-                         execution_manager=execution_manager)
+                         execution_manager=execution_manager,
+                         sim_model=sim_model)
         self._device_info: Mapping[str, str] = {}
         if temp_locks.get(port):
             self._driver = temp_locks[port][1]
         else:
-            self._driver = self._build_driver(simulating)  # type: ignore
+            self._driver = self._build_driver(  # type: ignore
+                simulating, sim_model or self.model())
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/modules/tempdeck.py]: `self.model` is based on device info and therefore won't be present yet, so this is a correct type check failure. `build_driver` should allow `None` for its model and we shouldn't do `sim_model or self.model()`, just `sim_model` (same for the other modules)

File: api/src/opentrons/hardware_control/modules/tempdeck.py [hunk 4]
```diff
@@ -142,26 +90,30 @@ def bootloader(cls) -> mod_abc.UploadFunction:
...
+                 sim_model: str = None) -> None:
         super().__init__(port=port,
                          simulating=simulating,
                          loop=loop,
-                         execution_manager=execution_manager)
+                         execution_manager=execution_manager,
+                         sim_model=sim_model)
         self._device_info: Mapping[str, str] = {}
         if temp_locks.get(port):
             self._driver = temp_locks[port][1]
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/modules/tempdeck.py]: if we annotate `self._driver` as `Union[ThermocyclerDriver, SimulatingDriver` we can get rid of the `type: ignore` (same for the other modules)

File: api/src/opentrons/protocol_api/protocol_context.py [hunk 5]
```diff
@@ -452,7 +452,9 @@ def load_module(
                     simulating=True,
                     loop=self._hw_manager.hardware.loop,
                     execution_manager=ExecutionManager(
-                        loop=self._hw_manager.hardware.loop)))
+                        loop=self._hw_manager.hardware.loop),
+                    sim_model=resolved_model.value))
+            hc_mod_instance._connect()
```
Comment (sfoster1) [inline api/src/opentrons/protocol_api/protocol_context.py]: this is nice!

File: api/tests/opentrons/api/test_session.py [hunk 6]
```diff
@@ -343,6 +343,7 @@ def run(ctx):
     mod = ctx.load_module('magdeck', '4')
     mod2 = ctx.load_module('tempdeck', '5')
     plate = mod2.load_labware('corning_96_wellplate_360ul_flat')
+    print(type(mod2))
```
Comment (sfoster1) [inline api/tests/opentrons/api/test_session.py]: diagnostic print, remove

**Review by sfoster1 — CHANGES_REQUESTED** [review]
Let's change some stuff to get rid of the `type: ignore` in driver assignments so typechecking continues to work, but the structure looks good

**Review by sfoster1 — APPROVED** [review]
LGTM if qa pass

**Review by nusrat813 — APPROVED** [review]
Tested - working as intended

**Pull request merged**